synapse/storage/schema/main/delta/65/06_msc2409_add_device_id_appservice_stream_type.sql

```diff
@@ -0,0 +1,18 @@
+/* Copyright 2021 The Matrix.org Foundation C.I.C
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+-- Add a column to track what to_device stream id that this application
+-- service has been caught up to.
+ALTER TABLE application_services_state ADD COLUMN to_device_stream_id BIGINT;
```

Comment:
could this be NOT NULL?

Reply:
> The rest of the columns in application_services_state are currently nullable:

yes, but synapse's schema is generally messy. We shouldn't set too much store by what happened in the past.

> For instance, we plan to not send historical to-device messages to an application service when it's first connected. 

hrm, this sounds like it might be a useful use of NULL here then. Has this code landed yet? I don't remember seeing it.

But if you're assigning a special meaning to NULL, you need to document it in the schema definition.

> On the other hand, what benefits would NOT NULL bring?

Mostly the same benefits of not using a TEXT field: less storage overhead, less having to worry about what to do when the data doesn't match your expected format.  Let the database do the validation for us.

Reply:
> hrm, this sounds like it might be a useful use of NULL here then. Has this code landed yet? I don't remember seeing it.

It hasn't landed, no. From the PR description:

> - [ ] When an application service is added for the first time to the homeserver, do we want it to receive historical to-device messages for the users in its namespace?
>    * TravisR says no, and I agree. The AS should still receive missed messages if it was down, but that's part of the later reliability step.
>    * **This will be done in a subsequent PR.**

Adding the meaning as a comment to the schema definition is a good idea though. I hadn't thought of that. I can do that here while we're adding the migration.

Reply:
Added the comment in 7cf6ad9197.